website: First draft of "Refactoring" documentation, about "moved" blocks

apparentlymart opened the pull request:
This is an early iteration of documentation about the refactoring features we're intending to introduce in a forthcoming release.

I've named the documentation section "refactoring" because in previous discussions with users that seems to be the term they use to describe the underlying need. `moved` blocks are our first language feature intended to meet that need, although it probably won't be the last as we consider other requirements in later releases. My intent here is that once we've published this it should eventually end up being the first result for a web search for the topic of Terraform refactoring.

ufukty commented on website/docs/language/modules/refactoring.html.md:
```diff
@@ -0,0 +1,468 @@
+
+## `moved` Block Syntax
+
+A `moved` block expects no labels and contains only `from` and `two` arguments:
```
typo

apparentlymart replied:
Thanks!

laurapacilio commented:
Hello! 👋  

I'm going to do a more thorough review of the drafted content, but one thing for us to consider as part of this PR is whether there are other pages within the docs where we'd want to link to or reference this new information. For example: Could we add a bullet on the modules development overview page? What about on the module blocks page (just a bit of context and a pointer to the refactoring page)? 

Also - I'm wondering about this article's position in the hierarchy. Is this something that module creators would need to do within their module's configuration so that their changes don't affect folks who are calling their module? or is this something that folks who are calling an updated/moved child module would need to do in their configuration? Both? The reason I ask is - Should this be housed in the module development section instead?  We can always do good cross-linking, wherever the page ends up being located, but wanted to put this out there.

apparentlymart commented:
Thanks for those initial thoughts, @laurapacilio!

I honestly hadn't noticed that we'd broken out "Module Development" into a separate category here, and now I know about it I agree that it makes more sense for this content to live there. As you suspected, this is primarily a feature for module developers to do refactoring without disrupting their callers.

Folks who are only _using_ modules will only interact with it in the limited sense that the Terraform plan UI will include some mention that an object was moved, which might call for us having some other documentation _somewhere_ to talk about that from the caller's perspective, but I'm optimistically hoping that the UI will be intuitive enough for most folks that we won't need to spend too many words on it in the docs.

I think our overall distinction between module usage and module development is still a bit shaky here, as a result of us only recently starting to formalize that distinction. For that reason I feel unsure where would be best to link to this page; it seems like a tricky one because it's something that many folks will never actually need to know about, and even once they do I'm not sure they'd _expect_ to find a language feature for it, and so I'd expected they'd be more likely to use a search engine for a general term like "terraform refactoring" and then hopefully turn this up as the first result. With that said, the overview of the Module Development section (currently titled [Creating Modules](https://www.terraform.io/docs/language/modules/develop/index.html)) seems like a short enough page that it wouldn't hurt to add a little about refactoring to it, perhaps as an additional paragraph under the "Module structure" section.

I'm going to be away from work for a little while over the next couple weeks so I expect I won't have time to do revisions here until I get back, but I agree with all of this feedback so far and do intend to incorporate it. Thanks!

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+
+## Renaming a Resource
+
+The following is an expanded version of the example above showing the full
+configurations at each step.
```
We probably don't need to say this - we can just launch in!

```suggestion
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+it covers both `aws_instance.a[0]` and `aws_instance.a[1]` without the need
+to identify each one separately.
+
+Each resource type has a separate schema and so objects of different types
+are not compatible. Therefore although you can use `moved` to change the name
+of a resource, you _cannot_ use `moved` to change to a different resource type
+or to change a managed resource (a `resource` block) into a data resource
+(a `data` block).
```
```suggestion
Each resource type has a separate schema and so objects of different types
are not compatible. Therefore, although you can use `moved` to change the name
of a resource, you _cannot_ use `moved` to change to a different resource type
or to change a managed resource (a `resource` block) into a data resource
(a `data` block).
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+it. In that case, the provider developer is responsible for making the old
+and new names schema-compatible, and can then tell Terraform that it's safe to
+support changing from the old to the new type via a `moved` block. Provider
+documentation will include an example for any situation where this exception
+applies.
```
are we sure there will always be an example? I wonder if we should soften and say "should" or do we enforce this somehow?

apparentlymart replied:
The mechanism I've described here doesn't exist yet today, so I wrote this with the intent that writing such documentation would be part of the work of making use of this new mechanism in providers.

However, it's true that this is written under the assumption that all such changes will be made by folks at HashiCorp who would therefore follow our documentation conventions/standards. We can't force third-party providers to have such documentation, so it is perhaps too strong to say "will".

With that said, there isn't really any way for a user to discover that this is possible _other than_ in documentation, so an undocumented instance of this would honestly be rather useless. I'm happy to change the wording as you suggested, but this is also a good note that we should make sure that the external provider-developer-facing documentation for this feature should mention this assumption too, once we have this facility available in the plugin SDK for provider developers to use. (This is actually not part of our initial scope of work here, so I could also just remove this statement altogether for now and restore it later once there's at least one resource type it's true for.)

apparentlymart replied:
After re-reading the full section to recover the context, I decided to just remove this paragraph altogether for now.

We can add more words about this once this mechanism is implemented and so it'll presumably be clearer to us what provider developers will be expected to do in order to enable it.

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+There is one exception to that constraint: a provider developer may mark a
+particular resource type name as deprecated and nominate a new name to replace
+it. In that case, the provider developer is responsible for making the old
+and new names schema-compatible, and can then tell Terraform that it's safe to
```
```suggestion
and new names schema-compatible, and can then tell Terraform that it is safe to
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+Applying this configuration would lead to Terraform creating an object
+bound to the address `aws_instance.a`.
+
+Later requirements might lead to you needing to use
+[`for_each`](../meta-arguments/for_each.html) with this resource, in order
+to systematically declare multiple instances. To preserve an object that
```
```suggestion
Later, you need to use [`for_each`](../meta-arguments/for_each.html) with this resource to systematically declare multiple instances. To preserve an object that
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+}
+
+# The old configuration used "count" and the new configuration
+# uses either "count" nor "for_each", and we want to keep
+# only the object at index 2.
```
should you change it to "or" rather than "nor"? OR should be be neither? 

Also, is there a way to rephrase without saying "we". Maybe we just change to "you"

apparentlymart replied:
Hmm yes, I intended this to be "neither". Thanks!

Regarding the "we"/"you" thing, I think historically we've typically used "we" in situations where we're leading someone through a hypothetical example, and "you" for when we're talking about something the reader _actually_ wants, though I'm probably overstating how consistent we've been in that way.

For this particular situation it feels weird to me to _tell_ the reader what they want, but maybe that's okay. Another option that I think of considering your feedback here would be to phrase it as an "If you want ... then ..." sentence, perhaps like this:

```hcl
# The old configuration used "count" and the new configuration
# uses either "count" nor "for_each". If you want to keep the
# object at index 2, declare that it has moved to the new
# address.
moved {
  from = aws_instance.d[2]
  to   = aws_instance.d
}
```

Does that seem reasonable?

(I guess I'd also need to similarly edit the other examples above, because they are all currently phrased as descriptions of a situation where the example would apply, rather than as an instruction for the reader to follow, but I'll wait until I hear what you think about _this_ example before I attempt that! :grinning: )

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,475 @@
+}
+```
+
+Imagine that we intend to split this into two modules as follows:
````
```suggestion
You can split this into two modules as follows:
```

apparentlymart replied:
One thing I didn't really like about my first draft was the various places where I wrote "Imagine that...", or similar. That's not a sentence structure we typically use, but I was struggling for some other way to introduce a hypothetical situation before showing how to address that situation.

I know your feedback here is about this specific case, and this one is a little different than the other examples anyway, but I'd love to hear if you have any ideas on how to restructure the other "Imagine that..." sentences here.

laurapacilio replied:
Thank you for asking! I added in some additional comments to try to address this. Generally I think yes - we don't want to say "imagine" though I completely understand why you initially did that. It can be hard to do this without phrases like "Let's say" or "Imagine that". But, I think actually that it's just fine to say "Consider this example module:" [show example] "Later, you split this into two modules and add a block....."   I think this way you're making it active, concise, and still walking the user through the example in a way that makes sense. I think this also works because before this section you tell users that they're about to see a bunch of example walkthroughs. THoughts?

apparentlymart replied:
Thanks! I've incorporated all of your suggestions below about "Imagine that..." sentences, and I previously incorporated the one I replied to here, so I'm going to resolve this conversation to represent that.

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+that call had not been a [local path](sources.html#local-paths).
+
+Module references in `moved` blocks are resolved relative to the module
+instance they are defined in. For example, if our original module above were
```
```suggestion
instance they are defined in. For example, if the original module above were
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+`terraform apply` with your new module version already then it can be safe
+to remove `moved` blocks in later versions of your module.
+
+However, in the general case removing a `moved` block is a breaking change
+to your module, because any configurations whose state refers to the old
+address will then plan to delete that existing object instead of move it.
```
```suggestion
However, removing a `moved` block is a generally breaking change
to your module because any configurations whose state refers to the old
address will then plan to delete that existing object instead of move it.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,475 @@
+}
+```
+
+When you record a sequence of moves in this way you will allow for successful
+upgrades both for configurations with objects at `aws_instance.a` _and_
+configurations with objects at `aws_instance.b`. In both cases, Terraform will
+treat the existing object as if it had been originally created as
+`aws_instance.c`.
````
```suggestion
Recording a sequence of moves in this way allows for successful
upgrades for both configurations with objects at `aws_instance.a` _and_
configurations with objects at `aws_instance.b`. In both cases, Terraform will
treat the existing object as if it had been originally created as
`aws_instance.c`.
```

laurapacilio commented on website/docs/language/modules/syntax.html.md:
```diff
@@ -160,37 +160,37 @@ For more information about referring to named values, see
+
+## Replacing resources within a module
+
+If you have an object that needs to be replaced with a new object for a reason
+that isn't automatically visible to Terraform, such as if a particular virtual
+machine is running on degraded underlying hardware, you can use
+[the `-replace=...` planning option](https://www.terraform.io/docs/cli/commands/plan.html#replace-address)
+to force Terraform to propose replacing that object.
```
```suggestion
You may have an object that needs to be replaced with a new object for a reason
that isn't automatically visible to Terraform, such as if a particular virtual
machine is running on degraded underlying hardware. In this case, you can use
[the `-replace=...` planning option](https://www.terraform.io/docs/cli/commands/plan.html#replace-address)
to force Terraform to propose replacing that object.
```

laurapacilio reviewed (approved):
This is really great Martin! I know you're still working on this, but I made a couple more nit suggestions for flow.

apparentlymart commented on website/docs/language/modules/develop/index.html.md:
```diff
@@ -70,3 +70,15 @@ your module is not creating any new abstraction and so the module is
 
+## Refactoring module resources
+
+In shared modules and long-lived configurations you may find that you outgrow
+your initial resource names and module organization and want to restructure
+them.
```
Just want to confirm that your intent here was to remove this paragraph altogether? Fine if so -- I was ambivalent myself as to whether this was just redundant with the stuff on the main Refactoring page anyway -- but I know the code suggestions mechanism can be kinda awkward for reviewing writing so want to make sure don't misunderstand!

laurapacilio replied:
Yes! Okay I'm so sorry - I should have actually explained myself here. My reasoning was just that this feels a bit redundant with the title and the first sentence of the paragraph after it. And so I was thinking we could safely cut it to make the text shorter. If you object, I respect that! But that was my thinking :-)

apparentlymart commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+understand moving or renaming an object as an intent to destroy the object
+at the old address and to create a new object at the new address.
+
+You can use `moved` blocks in your configuration to record where you've
+historically moved or renamed an object, in which case Terraform will instead
+treat an existing object at the old address as if it now belongs to the new
+address.
```
Good point!

I think the "why is this good" can vary between situations, so I'll have a think about how to write this in a more concise way, but in the interests of capturing the answer somewhere we can refer to it for later edits, reasons might include:

* As you suggested, it could be time-consuming to destroy and recreate some kinds of objects.
* The brief period where there are fewer (or no) objects of a particular type, or in some cases _too many_ objects of a particular type (if we're using `create_before_destroy`) can potentially cause total downtime or problematic backpressure on other subsystems. For example, adding new members to a Consul cluster will cause the cluster to temporarily thrash as it replicates data from the existing servers to the new servers, which might hurt performance for systems depending on that cluster.
* Some objects are not trivially replaceable. For example, if we have a resource representing a database server and we replace the database server then we'll probably be replacing a server which has data in it with a server that has _no_ data in it.

Perhaps the general way to talk about this is the idea that there are lots of infrastructure object types where replacing them requires careful planning and possibly coordination with other teams, and so module users and authors would typically prefer that _just_ upgrading to a new version of a module (without making other changes) would not involve replacing any existing objects.

I'll continue to ponder this, though if what I've said here gives _you_ any ideas about how to word this then of course I'd love to hear them!

laurapacilio replied:
I will think about this, too! Thank you as always for the awesome explanation :-)

apparentlymart commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,475 @@
+planning actions. The resulting plan is as if the object had originally been
+created at `aws_instance.b`, avoiding any need to destroy that object.
+
+The `from` and `to` addresses both use a special addressing syntax, different
+than other references in a module, which allows selecting both modules and
+resources, and possibly selecting resources inside child modules. The following
+sections describe the various different refactoring use-cases, along with the
+appropriate addressing syntax for each situation.
```
Agreed that the "possibly" doesn't really seem to be adding anything to this sentence, in retrospect. I think I was trying to say that the module path is an optional part of the address, but that's not really an important detail at this early point when we're just exploring the high-level structure.

The specific examples later hopefully serve to show the various different valid permutations.

apparentlymart commented:
I've incorporated most of the feedback from above exactly as written. Thanks!

I resolved all of the ones I just incorporated directly so it's easier to see what's left; I had a few questions about how to proceed with some of the less obvious situations.

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+
+## Enabling `count` or `for_each` For a Resource
+
+Imagine that the initial version of your module contains a single-instance
+resource:
```
```suggestion
Here is an example module that contains a single-instance
resource:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+Applying this configuration would lead to Terraform creating an object
+bound to the address `aws_instance.a`.
+
+Later, you need to use [`for_each`](../../meta-arguments/for_each.html) with this
+resource to systematically declare multiple instances. To preserve an object
+that was previously associated with `aws_instance.a` alone, you must use a
+`moved` block to specify which instance key the object will take in the new
+configuration:
```
```suggestion
In a later version, you use [`for_each`](../../meta-arguments/for_each.html) with this
resource to systematically declare multiple instances. To preserve an object
that was previously associated with `aws_instance.a` alone, you must add a
`moved` block to specify which instance key the object will take in the new
configuration:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+
+## Renaming a Resource
+
+Imagine that the inital version of your module included a resource configuration
+like the following:
```
```suggestion
Here is an example module with a resource configuration:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+Applying this configuration for the first time would cause Terraform to
+create `aws_instance.a[0]` and `aws_instance.a[1]`.
+
+If you later choose a better name for this resource, then you can change the
+name label in the `resource` block and record the old name inside a `moved` block:
```
```suggestion
If you choose a different name for this resource in a subsequent module version, you can change the
name label in the `resource` block and record the old name inside a `moved` block:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+
+## Enabling `count` or `for_each` For a Module Call
+
+Imagine that the initial version of your module contains a single-instance
+module:
```
```suggestion
Here is an example of a single-instance module:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+Applying this configuration would cause Terraform to create objects whose
+addresses begin with `module.a`.
+
+Later requirements might lead to you needing to use
+[`count`](../../meta-arguments/count.html) with this resource, in order
+to systematically declare multiple instances. To preserve an object that
+was previously associated with `aws_instance.a` alone, you can used a `moved`
+block to specify which instance key that object will take in the new
+configuration:
```
```suggestion
In later module versions, you may need to use [`count`](../../meta-arguments/count.html) with this resource to systematically declare multiple instances. To preserve an object that was previously associated with `aws_instance.a` alone, you can add a `moved` block to specify which instance key that object will take in the new configuration:
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+were originally created in `module.a[2]`, and thus it will plan to create
+new objects only for `module.a[0]` and `module.a[1]`.
+
+When at least one of the two addresses includes an instance key, like
+`[2]` in the above example, Terraform will understand both addresses as
+referring to specific _instances_ of a module call rather than the module
+call as a whole. That means you can use `moved` to switch between keys and to
+add and remove keys as you switch between `count`, `for_each`, or neither.
```
In general, we want to try to avoid clauses like this in the middle of the sentence. It's not _wrong_, it's just a bit harder to read than a more straightforward sentence construction. I'm removing this one here because I think no information is lost, but feel free to push back if you think it's necessary!

```suggestion
When at least one of the two addresses includes an instance key, Terraform will understand both addresses as
referring to specific _instances_ of a module call rather than the module
call as a whole. That means you can use `moved` to switch between keys and to
add and remove keys as you switch between `count`, `for_each`, or neither.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+As a module grows to support new requirements, it might eventually grow big
+enough to warrant splitting into two separate modules.
+
+Consider the following initial version of a module:
```
```suggestion
Consider this example module:
```

laurapacilio reviewed (approved):
Adding in a few more comments to address "imagine" language. I agree that this is a construction that we want to avoid. I think instead, the "Consider this example module" language is better. It's short, and the word "example" helps the user know that we're about to walk through something.  And then, I try to make things as clean and simple as possible as the scenario evolves. For example rather than "If later requirements need you to X" we can say "Later, you X".  Let me know if that doesn't make sense/if you want to talk about it more.

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+3. Edit the original module to no longer include any of these resources, and
+   instead to contain only shim configuration to migrate existing users.
+
+The new modules "x" and "y" will therefore contain only `resource` blocks,
+as normal:
```
One last thought - you use "as normal" a couple of times I think and I'm not 100% sure what it means. I wonder if you can just omit it as I've suggested here. 

```suggestion
The new modules "x" and "y" should contain only `resource` blocks:
```

apparentlymart replied:
With some time away from this _I'm_ honestly not sure what I meant by "as normal" here either, and I agree it doesn't really seem to be adding anything here.

(I _guess_ I was probably trying to say that there are no `moved` blocks in the child modules, because in this case they appear only in the root module, but another paragraph further down gets that message across more directly anyway. :man_shrugging: )

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+that what was previously one child module makes more sense as two separate
+modules and move a subset of the existing resources to the new one.
+
+Terraform compares previous state with new configuration, correlating by
+each module or resource's unique address. Therefore _by default_ Terraform will
+understand moving or renaming an object as an intent to destroy the object
+at the old address and to create a new object at the new address.
```
You write so well, Martin, that I'm nitpicking now :p But, I try to avoid saying "will." Often we can just make a stronger, simpler statement that means the same thing by just omitting it. 

```suggestion
Terraform compares previous state with new configuration, correlating by each module or resource's unique address. Therefore _by default_ Terraform understands moving or renaming an object as an intent to destroy the object at the old address and create a new object at the new address.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+understand moving or renaming an object as an intent to destroy the object
+at the old address and to create a new object at the new address.
+
+When you add `moved` blocks in your configuration to record where you've
+historically moved or renamed an object, Terraform will instead treat an
+existing object at the old address as if it now belongs to the new address.
```
```suggestion
When you add `moved` blocks in your configuration to record where you've
historically moved or renamed an object, Terraform treats an
existing object at the old address as if it now belongs to the new address.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+as if the object had originally been created at `aws_instance.b`, avoiding any
+need to destroy it during apply.
+
+The `from` and `to` addresses both use a special addressing syntax that allows
+selecting modules, resources, and resources inside child modules. Below, we
+describe the various different refactoring use-cases and the appropriate
+addressing syntax for each situation.
```
```suggestion
The `from` and `to` addresses both use a special addressing syntax that allows
selecting modules, resources, and resources inside child modules. Below, we
describe several refactoring use-cases and the appropriate
addressing syntax for each situation.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+When creating the next plan for each configuration using this module, Terraform
+will treat any existing objects belonging to `aws_instance.a` as if they had
+been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as
+`aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.
````
```suggestion
When creating the next plan for each configuration using this module, Terraform treats any existing objects belonging to `aws_instance.a` as if they had been created as `aws_instance.b`. `aws_instance.a[0]` will be treated as `aws_instance.b[0]`, and `aws_instance.a[1]` as `aws_instance.b[1]`.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+The above will avoid Terraform planning to destroy any existing object at
+`aws_instance.a`, treating it instead as if it were originally created
+as `aws_instance.a["small"]`.
````
```suggestion
The above will keep Terraform from planning to destroy any existing object at
`aws_instance.a`, treating those objects instead as if they were originally created
as `aws_instance.a["small"]`.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+`aws_instance.a`, treating it instead as if it were originally created
+as `aws_instance.a["small"]`.
+
+When at least one of the two addresses includes an instance key, like
+`["small"]` in the above example, Terraform will understand both addresses
```
```suggestion
When at least one of the two addresses includes an instance key, like
`["small"]` in the above example, Terraform understands both addresses
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+  to   = aws_instance.c["tiny"]
+}
+
+# The old configuration used "count" and the new configuration
+# uses neither "count" nor "for_each", and we want to keep
```
```suggestion
# The old configuration used "count" and the new configuration
# uses neither "count" nor "for_each", and you want to keep
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+The above will cause Terraform to treat all objects in `module.a` as if they
+were originally created in `module.a[2]`, and thus it will plan to create
+new objects only for `module.a[0]` and `module.a[1]`.
````
Going to keep the "will" here as we're describing something that is truly future from the example. 

```suggestion
The configuration above directs Terraform to treat all objects in `module.a` as if they were originally created in `module.a[2]`. As a result, Terraform plans to create new objects only for `module.a[0]` and 'module.a[1]`.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+
+You can split this into two modules as follows:
+
+* `aws_instance.a` will now belong to module "x".
+* `aws_instance.b` will also belong to module "x".
+* `aws_instance.c`, though, will belong to module "y".
```
```suggestion
* `aws_instance.a` now belongs to module "x".
* `aws_instance.b` also belongs to module "x".
* `aws_instance.c` belongs to module "y".
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+When an existing user of the original module upgrades to the new "shim"
+version, Terraform will notice these three `moved` blocks and thus behave
+as if the objects associated with the three old resource addresses were
+originally created inside the two new modules.
````
```suggestion
When an existing user of the original module upgrades to the new "shim"
version, Terraform notices these three `moved` blocks and behaves
as if the objects associated with the three old resource addresses were
originally created inside the two new modules.
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+and distributed together in a single
+[module package](../sources.html#modules-in-package-sub-directories).
+
+To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
+between separately-packaged modules, Terraform will only allow declarations of
+moves between modules in the same package. In other words, Terraform would
+not have allowed moving into `module.x` above if the `source` address of
+that call had not been a [local path](../sources.html#local-paths).
```
```suggestion
To avoid strong [coupling](https://en.wikipedia.org/wiki/Coupling_(computer_programming))
between separately-packaged modules, Terraform only allows declarations of
moves between modules in the same package. In other words, Terraform would
not have allowed moving into `module.x` above if the `source` address of
that call had not been a [local path](../sources.html#local-paths).
```

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
```diff
@@ -0,0 +1,461 @@
+
+Over time, a long-lasting module may accumulate many `moved` blocks.
+
+If you are maintaining private modules within an organization and you know
+for certain that all users of your module have successfully run
+`terraform apply` with your new module version already then it can be safe
+to remove `moved` blocks in later versions of your module.
```
Where possible, I try to rework sentences with a long "if" clause at the beginning. Flipping them around like this generally makes them easier to read :-) 

```suggestion
It can be safe to remove `moved` blocks in later versions of your module when you are maintaining private modules within an organization and all users of your module have successfully run
`terraform apply` with your new module version.
```

apparentlymart replied:
This restructuring looks good to me. Thanks!

One way your new paragraph feels different to my original intent is that I was using "know for certain" (admittedly, clumsily) to try to suggest that this is something that the person making the decision needs to actively check, rather than just to _assume_.

Perhaps that's not really so important as I originally thought, but the consequence of getting this wrong is that a user of the module that _hadn't_ yet run `terraform apply` with a version including the `moved` block will see a plan to _delete_ the object that was moved, and applying that plan could be hard to recover from if the object is something like a database or some similar server which has state embedded in it.

I'm not super set on including this, but I wonder if you have some thoughts about this.

apparentlymart replied:
When I was revising this I changed that part to say "...when you are maintaining private modules within an organization and _you know that_ all module users have...", but otherwise took this exactly as you wrote it.

I'm happy to still drop the extra words if you think they are unnecessary!

laurapacilio replied:
yeah that's fair - I actually went back and forth about that. It's minor and I think a little extra prompting never hurt anyone :-) I agree with what you did!

apparentlymart replied:
Thanks!

laurapacilio commented on website/docs/language/modules/develop/refactoring.html.md:
````diff
@@ -0,0 +1,461 @@
+}
+```
+
+Recording a sequence of moves in this way allows for successful upgrades for
+both configurations with objects at `aws_instance.a` _and_ configurations with
+objects at `aws_instance.b`. In both cases, Terraform will treat the existing
+object as if it had been originally created as `aws_instance.c`.
````
```suggestion
Recording a sequence of moves in this way allows for successful upgrades for
both configurations with objects at `aws_instance.a` _and_ configurations with
objects at `aws_instance.b`. In both cases, Terraform treats the existing
object as if it had been originally created as `aws_instance.c`.
```

laurapacilio commented on website/docs/language/modules/syntax.html.md:
````diff
@@ -160,37 +160,37 @@ For more information about referring to named values, see
-
-```shell
-$ terraform taint module.salt_master.aws_instance.salt_master
+When refactoring an existing configuration to split resources across multiple
+child modules, just moving the `resource` blocks into the child modules would
+cause Terraform to see the new location as an entirely different resource, and
+so Terraform will propose to destroy all resource instances at the old address
+and create new instances at the new address.
````
```suggestion
Moving `resource` blocks from one module into several child modules causes Terraform to see the new location as an entirely different resource. As a result, Terraform plans to destroy all resource instances at the old address and create new instances at the new address.
```

laurapacilio commented on website/docs/language/modules/syntax.html.md:
```diff
@@ -160,37 +160,37 @@ For more information about referring to named values, see
+so Terraform will propose to destroy all resource instances at the old address
+and create new instances at the new address.
+
+To preserve existing objects when refactoring between modules, you can use
+[refactoring blocks](develop/refactoring.html) to record the old and new addresses
+for each change you made. Terraform will then treat existing objects at the
+old addresses as if they had originally been created at the corresponding
+new addresses.
```
```suggestion
To preserve existing objects, you can use [refactoring blocks](develop/refactoring.html) to record the old and new addresses for each resource. This directs Terraform to treat existing objects at the
old addresses as if they had originally been created at the corresponding new addresses.
```

laurapacilio commented on website/docs/language/modules/syntax.html.md:
```diff
@@ -160,37 +160,37 @@ For more information about referring to named values, see
+[the `-replace=...` planning option](/docs/cli/commands/plan.html#replace-address)
+to force Terraform to propose replacing that object.
+
+If the object in question belongs to a resource within a nested module,
+specify the full path to that resource including all of the nested module
+steps leading to it. For example:
```
```suggestion
If the object belongs to a resource within a nested module,
specify the full path to that resource including all of the nested module
steps leading to it. For example:
```

laurapacilio commented on website/docs/language/modules/syntax.html.md:
```diff
@@ -160,37 +160,37 @@ For more information about referring to named values, see
 
-It is not possible to taint an entire module. Instead, each resource within
-the module must be tainted separately.
+The above would select a `resource "aws_instance" "example"` declared inside
+a `module "example"` child module declared inside your root module.
```
```suggestion
The above selects a `resource "aws_instance" "example"` declared inside
a `module "example"` child module declared inside your root module.
```

apparentlymart commented:
We do have a few little things dangling above but it feels to me like this is in reasonable shape for the first release of this feature. Would you agree, @laurapacilio?

I'm going to hold on merging this until the actual implementation is far enough along that we remove its experimental feature warnings in the UI, because it'd be confusing to publish this anywhere while the implementation work is still not complete, so we do have some time for further editing but unless you object I'm going to plan to merge this once the time is right, so we can make sure it's in place in time for releasing the feature.

Thanks again for all of the suggestions!

laurapacilio commented:
I do agree! Thanks for working with me on this - I definitely think it's ready for the first release. Great job, Martin - ship it when ready! 🚀

apparentlymart commented:
Unfortunately this website-link-check failure has been showing up in a bunch of places and isn't directly related to this changeset. It seems like the container image we use for that job might need updating to include a fresher set of CA certificates in order to actually trust the certs on the external sites we're linking to, but we'll need to tackle that elsewhere.

laurapacilio commented:
Yes! I believe folks are working on these :-) https://github.com/hashicorp/terraform-website/issues/1925

Don't let those be a blocker to merge this if you need to!
